# **Automatidata project**
**Course 2 - Get Started with Python**

Welcome to the Automatidata Project!

You have just started as a data professional in a fictional data consulting firm, Automatidata. Their client, the New York City Taxi and Limousine Commission (New York City TLC), has hired the Automatidata team for its reputation in helping their clients develop data-based solutions.

The team is still in the early stages of the project. Previously, you were asked to complete a project proposal by your supervisor, DeShawn Washington. You have received notice that your project proposal has been approved and that New York City TLC has given the Automatidata team access to their data. To get clear insights, New York TLC's data must be analyzed, key variables identified, and the dataset ensured it is ready for analysis.

A notebook was structured and prepared to help you in this project. Please complete the following questions.

# Course 2 End-of-course project: Inspect and analyze data

In this activity, you will examine data provided and prepare it for analysis.  This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>    

**The purpose** of this project is to investigate and understand the data provided.
  
**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings. 
<br/>  
*This activity has three parts:*

**Part 1:** Understand the situation 
* Prepare to understand and organize the provided taxi cab dataset and information.

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities.

* Compile summary information about the data to inform next steps.

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into specific variables.


<br/> 
Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work. 



# **Identify data types and relevant variables using Python**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## PACE: **Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided taxi cab information? 

To effectively prepare for understanding and organizing the provided taxi cab dataset from the New York City Taxi and Limousine Commission (TLC), I will follow a structured approach to ensure the data is ready for analysis and insight generation. Below is a summary of my preparation plan:

1. **Clarify Project Objectives**  
   I will begin by reviewing the project brief and confirming the primary business questions we aim to answer. This will help ensure that our analysis aligns with the expectations of the New York City TLC and focuses on actionable insights (e.g., trip patterns, fare analysis, and peak hour identification).

2. **Review Data Documentation**  
   If available, I will consult any accompanying data dictionaries or metadata to understand the structure and meaning of each variable in the dataset. This includes fields such as pickup/dropoff times, trip distance, fare amount, and passenger count.

3. **Set Up the Analysis Environment**  
   I will prepare a clean Python environment using key libraries such as `pandas`, `numpy`, and `matplotlib/seaborn` for analysis and visualization. The dataset will be loaded into a pandas DataFrame for inspection and manipulation.

4. **Conduct an Initial Data Review**  
   I will display the first few rows of the dataset using `.head()` and check for any immediate issues such as null values, incorrect data types, or outliers (e.g., trips with zero distance or excessively high fares).

5. **Plan for Data Cleaning and Transformation**  
   I will identify potential data quality issues, including missing or inconsistent values, incorrect formats (e.g., datetime), and duplicates. This will be addressed during the data preparation step to ensure reliability in analysis.

6. **Outline Analytical Objectives**  
   Based on the initial review, I will outline key metrics and potential visualizations that will be helpful for exploratory data analysis, such as distribution of trip distances, fare trends, and heatmaps of pickup locations.

By following this process, I will ensure that the dataset is well-understood, properly organized, and ready for deeper analysis in the next phases of the project.


<img src="images/Analyze.png" width="100" height="100" align=left>

## PACE: **Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Build dataframe**
















Create a pandas dataframe for data learning, and future exploratory data analysis (EDA) and statistical activities.

**Code the following,**

*   import pandas as `pd`. pandas is used for buidling dataframes.

*   import numpy as `np`. numpy is imported with pandas

*   `df = pd.read_csv('Datasets\NYC taxi data.csv')`

**Note:** pair the data object name `df` with pandas functions to manipulate data, such as `df.groupby()`.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [41]:
#Import libraries and packages listed above
import pandas as pd
import numpy as np

# Load dataset into dataframe
df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')
print(df)

       Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0        24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1        35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2       106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3        38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4        30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   
...           ...       ...                     ...                     ...   
22694    14873857         2   02/24/2017 5:37:23 PM   02/24/2017 5:40:39 PM   
22695    66632549         2   08/06/2017 4:43:59 PM   08/06/2017 5:24:47 PM   
22696    74239933         2   09/04/2017 2:54:14 PM   09/04/2017 2:58:22 PM   
22697    60217333         2  07/15/2017 12:56:30 PM   07/15/2017 1:08:26 PM   
22698    17208911         1   03/02/2017 1:02:49 PM   03/02/2017 1:16:09 PM   

       passenger_count  trip_distance  RatecodeID s

### **Task 2b. Understand the data - Inspect the data**

View and inspect summary information about the dataframe by coding the following:

1. `df.head(10)`
2. `df.info()`
3. `df.describe()`

Consider the following two questions:

**Question 1:** When reviewing the `df.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Question 2:** When reviewing the `df.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values?

### Question 1: When reviewing the `df.info()` output, what do you notice about the different variables?

Based on the `df.info()` output, I observed the following:

- The dataset contains **22,699 entries** across **18 columns**, and **no null values** are reported, which indicates data completeness at the structural level.
- There are **three object-type variables**:  
  - `tpep_pickup_datetime` and `tpep_dropoff_datetime`, which represent date/time values but are currently stored as strings and will need to be converted to datetime format.
  - `store_and_fwd_flag`, which is categorical (likely a yes/no flag indicating if the trip record was stored and forwarded).
- The remaining variables are either `int64` or `float64`, and are assumed to be **numeric and quantitative**.
- Column `Unnamed: 0` seems to be an index or ID column from the original CSV export and may be **redundant** for analysis.
- There are **no immediately visible data type mismatches** aside from the datetime fields needing conversion.

---

### Question 2: When reviewing the `df.describe()` output, what do you notice about the distributions of each variable?

Upon reviewing the descriptive statistics, several notable observations stand out:

- **Potential data quality issues:**
  - `passenger_count` has a **minimum value of 0**, which is not logical for a taxi trip and may indicate missing or misrecorded data.
  - `trip_distance` also has a **minimum value of 0**, which might indicate canceled trips, errors, or placeholder values.
  - The `fare_amount`, `extra`, `mta_tax`, `improvement_surcharge`, and `total_amount` columns all contain **negative values**, which are likely data entry errors or misclassifications.
  - `RatecodeID` ranges from 1 to 99, which may include **invalid or rare codes**, as standard NYC rate codes usually fall within a smaller range.
  - `tip_amount` has a maximum of 200, and `total_amount` a maximum of 1,200.29, which suggests the presence of **extreme outliers** worth further investigation.
- **Typical fare amounts (based on quartiles):**
  - The **median fare amount** is 9.50, with most fares falling between 6.50 and 14.50.
  - The **median trip distance** is approximately 1.61 miles, which aligns with typical short-distance urban taxi rides.

In summary, although the dataset appears structurally complete, several variables contain **anomalies or questionable values** that will need to be addressed during the data cleaning phase to ensure the reliability of any subsequent analysis.


In [42]:
df.head(10)

,Unnamed: 0,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,24870114,2,03/25/2017 8:55:43 AM,03/25/2017 9:09:47 AM,6,3.34,1,N,100,231,1,13.0,0.0,0.5,2.76,0.0,0.3,16.56
1,35634249,1,04/11/2017 2:53:28 PM,04/11/2017 3:19:58 PM,1,1.80,1,N,186,43,1,16.0,0.0,0.5,4.00,0.0,0.3,20.80
2,106203690,1,12/15/2017 7:26:56 AM,12/15/2017 7:34:08 AM,1,1.00,1,N,262,236,1,6.5,0.0,0.5,1.45,0.0,0.3,8.75
3,38942136,2,05/07/2017 1:17:59 PM,05/07/2017 1:48:14 PM,1,3.70,1,N,188,97,1,20.5,0.0,0.5,6.39,0.0,0.3,27.69
4,30841670,2,04/15/2017 11:32:20 PM,04/15/2017 11:49:03 PM,1,4.37,1,N,4,112,2,16.5,0.5,0.5,0.00,0.0,0.3,17.80
5,23345809,2,03/25/2017 8:34:11 PM,03/25/2017 8:42:11 PM,6,2.30,1,N,161,236,1,9.0,0.5,0.5,2.06,0.0,0.3,12.36
6,37660487,2,05/03/2017 7:04:09 PM,05/03/2017 8:03:47 PM,1,12.83,1,N,79,241,1,47.5,1.0,0.5,9.86,0.0,0.3,59.16
7,69059411,2,08/15/2017 5:41:06 PM,08/15/2017 6:03:05 PM,1,2.98,1,N,237,114,1,16.0,1.0,0.5,1.78,0.0,0.3,19.58
8,8433159,2,02/04/2017 4:17:07 PM,02/04/2017 4:29:14 PM,1,1.20,1,N,234,249,2,9.0,0.0,0.5,0.00,0.0,0.3,9.80
9,95294817,1,11/10/2017 3:20:29 PM,11/10/2017 3:40:55 PM,1,1.60,1,N,239,237,1,13.0,0.0,0.5,2.75,0.0,0.3,16.55


In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [44]:
df.describe()

,Unnamed: 0,VendorID,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
count,2.269900e+04,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000,22699.000000
mean,5.675849e+07,1.556236,1.642319,2.913313,1.043394,162.412353,161.527997,1.336887,13.026629,0.333275,0.497445,1.835781,0.312542,0.299551,16.310502
std,3.274493e+07,0.496838,1.285231,3.653171,0.708391,66.633373,70.139691,0.496211,13.243791,0.463097,0.039465,2.800626,1.399212,0.015673,16.097295
min,1.212700e+04,1.000000,0.000000,0.000000,1.000000,1.000000,1.000000,1.000000,-120.000000,-1.000000,-0.500000,0.000000,0.000000,-0.300000,-120.300000
25%,2.852056e+07,1.000000,1.000000,0.990000,1.000000,114.000000,112.000000,1.000000,6.500000,0.000000,0.500000,0.000000,0.000000,0.300000,8.750000
50%,5.673150e+07,2.000000,1.000000,1.610000,1.000000,162.000000,162.000000,1.000000,9.500000,0.000000,0.500000,1.350000,0.000000,0.300000,11.800000
75%,8.537452e+07,2.000000,2.000000,3.060000,1.000000,233.000000,233.000000,2.000000,14.500000,0.500000,0.500000,2.450000,0.000000,0.300000,17.800000
max,1.134863e+08,2.000000,6.000000,33.960000,99.000000,265.000000,265.000000,4.000000,999.990000,4.500000,0.500000,200.000000,19.100000,0.300000,1200.290000


### **Task 2c. Understand the Data - Investigate the variables**

Sort and interpret the data table for two variables:`trip_distance` and `total_amount`.

**Answer the following three questions:**

**Question 1:** Sort your first variable (`trip_distance`) from maximum to minimum value, do the values seem normal?

**Question 2:** Sort by your second variable (`total_amount`), are any values unusual?

**Question 3:** Are the resulting rows similar for both sorts? Why or why not?

**Question 1:**  
The values for `trip_distance` appear normal and reasonable, with the longest trips likely representing routes to or from airports or distant boroughs. No significant anomalies were observed in the distance values themselves.

**Question 2:**  
There are some unusual values in the `total_amount` column. For example, high fares associated with short distances and several negative or zero values indicate potential data entry errors or invalid records. These anomalies suggest the need for data cleaning before analysis.

**Question 3:**  
The rows resulting from sorting by `trip_distance` and `total_amount` are not the same, which is expected. This is because total fare is influenced by multiple components (e.g., tips, surcharges, tolls) and not just the distance traveled. Some high total amounts are unrelated to long trips, indicating possible inconsistencies or outliers in the data.


In [45]:
# Sort the data by trip distance from maximum to minimum value
df_sorted_by_distance = df.sort_values(by='trip_distance', ascending=False)
print(df_sorted_by_distance[['trip_distance', 'total_amount']].head())

       trip_distance  total_amount
9280           33.96        150.30
13861          33.92        258.21
6064           32.72        179.06
10291          31.95        131.80
29             30.83        111.38


In [46]:
# Sort the data by total amount and print the top 20 values
df_sorted_by_total_top = df.sort_values(by='total_amount', ascending=False)
print("\nTop 20 total_amount values:")
print(df_sorted_by_total_top[['total_amount', 'fare_amount', 'trip_distance']].head(20))


Top 20 total_amount values:
       total_amount  fare_amount  trip_distance
8476        1200.29       999.99           2.60
20312        450.30       450.00           0.00
13861        258.21       200.01          33.92
12511        233.74       175.00           0.00
15474        211.80       200.00           0.00
6064         179.06       107.00          32.72
16379        157.06       140.00          25.50
3582         152.30       152.00           7.30
11269        151.82       120.00           0.00
9280         150.30       150.00          33.96
1928         137.80       120.00          12.50
10291        131.80       131.00          31.95
6708         126.00       100.00           0.32
11608        123.30        99.50          23.00
908          121.56       100.00          26.12
7281         120.96       100.00           0.00
18130        119.31        90.50          30.50
13621        115.94       105.00          19.80
13359        111.95        75.00           0.00
29         

In [47]:
# Sort the data by total amount and print the bottom 20 values
df_sorted_by_total_bottom = df.sort_values(by='total_amount', ascending=True)
print("\nBottom 20 total_amount values:")
print(df_sorted_by_total_bottom[['total_amount', 'fare_amount', 'trip_distance']].head(20))


Bottom 20 total_amount values:
       total_amount  fare_amount  trip_distance
12944       -120.30      -120.00           0.17
20698         -5.80        -4.50           0.70
17602         -5.80        -4.00           0.46
11204         -5.30        -4.50           0.64
14714         -5.30        -4.00           0.40
8204          -4.80        -3.50           0.41
20317         -4.80        -3.50           0.24
10281         -4.30        -2.50           0.00
5448          -4.30        -3.50           0.25
4423          -4.30        -3.00           0.06
18565         -3.80        -3.00           0.10
314           -3.80        -2.50           0.12
5758          -3.80        -2.50           0.02
1646          -3.30        -2.50           0.04
10506          0.00         0.00           0.00
4402           0.00         0.00           7.06
5722           0.00         0.00           0.00
22566          0.00         0.00           0.00
19067          0.30         0.00           0.10
14283   

In [48]:
# Define a mapping dictionary for payment_type
payment_type_labels = {
    1: 'Credit card',
    2: 'Cash',
    3: 'No charge',
    4: 'Dispute',
    5: 'Unknown',
    6: 'Voided trip'
}

# Replace numeric codes with descriptive labels
df['payment_type'] = df['payment_type'].map(payment_type_labels)

# Count the number of each payment type
payment_type_counts = df['payment_type'].value_counts()
print("Counts of each payment type:")
print(payment_type_counts)

Counts of each payment type:
Credit card    15265
Cash            7267
No charge        121
Dispute           46
Name: payment_type, dtype: int64


According to the data dictionary, the payment method was encoded as follows:

1 = Credit card  
2 = Cash  
3 = No charge  
4 = Dispute  
5 = Unknown  
6 = Voided trip

In [49]:
# What is the average tip for trips paid for with credit card?
df[df['payment_type'] == 'Credit card']['tip_amount'].mean()


2.7298001965279934

In [50]:
# What is the average tip for trips paid for with cash?
df[df['payment_type'] == 'Cash']['tip_amount'].mean()


0.0

In [51]:
# How many times is each vendor ID represented in the data?
df['VendorID'].value_counts()



2    12626
1    10073
Name: VendorID, dtype: int64

In [52]:
# What is the mean total amount for each vendor?
df.groupby('VendorID')['total_amount'].mean()



VendorID
1    16.298119
2    16.320382
Name: total_amount, dtype: float64

In [53]:
# Filter the data for credit card payments only
cc_df = df[df['payment_type'] == 'Credit card']
cc_df

,Unnamed: 0,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,24870114,2,03/25/2017 8:55:43 AM,03/25/2017 9:09:47 AM,6,3.34,1,N,100,231,Credit card,13.0,0.0,0.5,2.76,0.00,0.3,16.56
1,35634249,1,04/11/2017 2:53:28 PM,04/11/2017 3:19:58 PM,1,1.80,1,N,186,43,Credit card,16.0,0.0,0.5,4.00,0.00,0.3,20.80
2,106203690,1,12/15/2017 7:26:56 AM,12/15/2017 7:34:08 AM,1,1.00,1,N,262,236,Credit card,6.5,0.0,0.5,1.45,0.00,0.3,8.75
3,38942136,2,05/07/2017 1:17:59 PM,05/07/2017 1:48:14 PM,1,3.70,1,N,188,97,Credit card,20.5,0.0,0.5,6.39,0.00,0.3,27.69
5,23345809,2,03/25/2017 8:34:11 PM,03/25/2017 8:42:11 PM,6,2.30,1,N,161,236,Credit card,9.0,0.5,0.5,2.06,0.00,0.3,12.36
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
22692,60425673,1,07/16/2017 3:22:51 AM,07/16/2017 3:40:52 AM,1,5.70,1,N,249,17,Credit card,19.0,0.5,0.5,4.05,0.00,0.3,24.35
22693,67858616,2,08/10/2017 10:20:04 PM,08/10/2017 10:29:31 PM,1,0.89,1,N,229,170,Credit card,7.5,0.5,0.5,1.76,0.00,0.3,10.56
22695,66632549,2,08/06/2017 4:43:59 PM,08/06/2017 5:24:47 PM,1,16.71,2,N,132,164,Credit card,52.0,0.0,0.5,14.64,5.76,0.3,73.20
22697,60217333,2,07/15/2017 12:56:30 PM,07/15/2017 1:08:26 PM,1,2.36,1,N,68,144,Credit card,10.5,0.0,0.5,1.70,0.00,0.3,13.00


In [54]:
# Filter the credit-card-only data for passenger count only
cc_df_passenger = cc_df[['passenger_count', 'tip_amount']]
cc_df_passenger

,passenger_count,tip_amount
0,6,2.76
1,1,4.00
2,1,1.45
3,1,6.39
5,6,2.06
...,...,...
22692,1,4.05
22693,1,1.76
22695,1,14.64
22697,1,1.70


In [55]:
# Calculate the average tip amount for each passenger count (credit card payments only)
cc_df_passenger.groupby('passenger_count')['tip_amount'].mean()

passenger_count
0    2.610370
1    2.714681
2    2.829949
3    2.726800
4    2.607753
5    2.762645
6    2.643326
Name: tip_amount, dtype: float64

<img src="images/Construct.png" width="100" height="100" align=left>

## PACE: **Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project. 




<img src="images/Execute.png" width="100" height="100" align=left>

## PACE: **Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response.


### **Given your efforts, what can you summarize for DeShawn and the data team?**

*Note for Learners: Your notebook should contain data that can address Luana's requests. Which two variables are most helpful for building a predictive model for the client: NYC TLC?*

###  Summary for DeShawn and the Data Team

Based on the analysis conducted, here are the key takeaways:

- **Average Tip by Payment Type**:
  - Trips paid by **credit card** had an average tip of **2.73**.
  - Trips paid by **cash** showed an average tip of **0.0**, which aligns with the data dictionary: *“cash tips are not included.”*

- **Vendor Distribution and Spending**:
  - The majority of rides were provided by:
    - **VeriFone Inc.**: 12,626 trips  
    - **Creative Mobile Technologies, LLC**: 10,073 trips
  - The **average total amount** per ride is:
    - Creative Mobile Technologies, LLC: **16.30**  
    - VeriFone Inc.: **16.32**

- **Passenger Count vs. Tip Amount (Credit Card Only)**:
  - There is minimal variation in average tip amounts across different passenger counts.
  - Tips tend to slightly increase with higher passenger counts, but the differences are not substantial.
    - 1 passenger: **2.71**
    - 6 passengers: **2.64**

---

###  Recommendation for Predictive Modeling (for NYC TLC)

The two variables that would be most helpful for building a predictive model are:

1. **Trip Distance (`trip_distance`)** – a continuous variable directly related to fare and total amount.
2. **Fare Amount (`fare_amount`)** – a strong numerical predictor tied to both trip distance and tip behavior, especially for card payments.

These variables can help NYC TLC forecast expected revenue, detect outliers, and monitor trends in ride behavior.


**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.